volga/streaming/runtime/network/transfer/local/data_reader.py

```diff
@@ -0,0 +1,132 @@
+import logging
+import time
+from collections import deque
+from typing import List, Optional
+
+import zmq
+
+from volga.streaming.runtime.network.buffer.buffer import get_payload, get_buffer_id, AckMessage, AckMessageBatch
+from volga.streaming.runtime.network.stats import Stats, StatsEvent
+from volga.streaming.runtime.network.transfer.io_loop import Direction
+from volga.streaming.runtime.network.channel import Channel, ChannelMessage
+
+import simplejson
+
+from volga.streaming.runtime.network.config import NetworkConfig, DEFAULT_NETWORK_CONFIG
+from volga.streaming.runtime.network.transfer.local.local_data_handler import LocalDataHandler
+from volga.streaming.runtime.network.utils import bytes_to_str, rcv_no_block, send_no_block
+
+logger = logging.getLogger("ray")
+
+
+class DataReader(LocalDataHandler):
+
+    def __init__(
+        self,
+        name: str,
+        channels: List[Channel],
+        node_id: str,
+        zmq_ctx: zmq.Context,
+        network_config: NetworkConfig = DEFAULT_NETWORK_CONFIG
+    ):
+        super().__init__(
+            name=name,
+            channels=channels,
+            node_id=node_id,
+            zmq_ctx=zmq_ctx,
+            direction=Direction.RECEIVER,
+            network_config=network_config
+        )
+
+        self.stats = Stats()
+
+        self._output_queue = deque()
+
+        self._watermarks = {c.channel_id: -1 for c in self._channels}
+        self._out_of_order = {c.channel_id: {} for c in self._channels}
+        self._acks_queues = {c.channel_id: deque() for c in self._channels}
+
+    def read_message(self) -> List[ChannelMessage]:
+        if len(self._output_queue) == 0:
+            return []
+        buffer = self._output_queue.popleft()
+        payload = get_payload(buffer)
+
+        # TODO implement partial messages
+        res = []
+        for (msg_id, data) in payload:
+            msg = simplejson.loads(bytes_to_str(data))
+            res.append(msg)
+        return res
+
+    def send(self, socket: zmq.Socket):
```

Comment:
Seems to be not so obvious that `reader` has `send` method. When I looked into the method, it appears to send only `acks`, so maybe we can do something about it

Reply:
This is because `DataReader` (as well as `DataWriter`) both implement `IOHandler` interface which has both `send() `and `rcv().` Not sure how properly name here to avoid confusion